File: 202-happy-number/202-happy-number.cpp
class Solution {
public:
    int fun( int x )
    {
        int total = 0;
        while( x != 0 ) {
            total += ( x % 10 ) * ( x % 10 );
            x = x / 10;    
        }
            
        return total;
    }
        
    bool isHappy(int n) {
       int slow = n;
       int fast = n;
       while ( 1 ) {
           slow = fun( slow );
           fast = fun( fun ( fast ) ) ;
           if( slow == fast ) {
               return (fast == 1);
           }
      } 
    }
};

File: 202-happy-number/202-happy-number.py
class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        
        if n == 1:
            return True
        
        def f(x):
            result  = 0
            
            while x != 0:
                result += (x % 10) ** 2
                x = x / 10
            return result
        
        fast = n
        slow = n
        
        while True:
            fast = f(f(fast))
            slow = f(slow)
            
            if fast == 1:
                return True
            elif slow == fast:
                return False
            
            
            
            
            
            
            
        
        